data/transactions/application.go

```diff
@@ -0,0 +1,769 @@
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with go-algorand.  If not, see <https://www.gnu.org/licenses/>.
+
+package transactions
+
+import (
+	"fmt"
+
+	"github.com/algorand/go-algorand/data/basics"
+)
+
+// OnCompletion is an enum representing some layer 1 side effect that an
+// ApplicationCall transaction will have if it is included in a block.
+type OnCompletion uint64
+
+const (
+	// NoOpOC indicates that an application transaction will simply call its
+	// ApprovalProgram
+	NoOpOC OnCompletion = 0
+
+	// OptInOC indicates that an application transaction will allocate some
+	// LocalState for the application in the sender's account
+	OptInOC OnCompletion = 1
+
+	// CloseOutOC indicates that an application transaction will deallocate
+	// some LocalState for the application from the user's account
+	CloseOutOC OnCompletion = 2
+
+	// ClearStateOC is similar to CloseOutOC, but may never fail. This
+	// allows users to reclaim their minimum balance from an application
+	// they no longer wish to opt in to.
+	ClearStateOC OnCompletion = 3
+
+	// UpdateApplicationOC indicates that an application transaction will
+	// update the ApprovalProgram and ClearStateProgram for the application
+	UpdateApplicationOC OnCompletion = 4
+
+	// DeleteApplicationOC indicates that an application transaction will
+	// delete the AppParams for the application from the creator's balance
+	// record
+	DeleteApplicationOC OnCompletion = 5
+)
+
+func (oc OnCompletion) String() string {
+	switch oc {
+	case NoOpOC:
+		return "noop"
+	case OptInOC:
+		return "optin"
+	case CloseOutOC:
+		return "closeout"
+	case ClearStateOC:
+		return "clearstate"
+	case UpdateApplicationOC:
+		return "update"
+	case DeleteApplicationOC:
+		return "delete"
+	}
+	return "?"
+}
+
+// ApplicationCallTxnFields captures the transaction fields used for all
+// interactions with applications
+type ApplicationCallTxnFields struct {
+	_struct struct{} `codec:",omitempty,omitemptyarray"`
+
+	ApplicationID   basics.AppIndex   `codec:"apid"`
+	OnCompletion    OnCompletion      `codec:"apan"`
+	ApplicationArgs [][]byte          `codec:"apaa,allocbound=256"`
+	Accounts        []basics.Address  `codec:"apat,allocbound=256"`
+	ForeignApps     []basics.AppIndex `codec:"apfa,allocbound=256"`
+
+	LocalStateSchema  basics.StateSchema `codec:"apls"`
+	GlobalStateSchema basics.StateSchema `codec:"apgs"`
+	ApprovalProgram   []byte             `codec:"apap"`
+	ClearStateProgram []byte             `codec:"apsu"`
+
+	// If you add any fields here, remember you MUST modify the Empty
+	// method below!
+}
+
+// Empty indicates whether or not all the fields in the
+// ApplicationCallTxnFields are zeroed out
+func (ac *ApplicationCallTxnFields) Empty() bool {
```

Comment:
Is there an easy way to use the `msgp` `.Zero` version of this?

Reply:
Also, since this is pass-by-pointer, consider checking for `ac == nil` here.

Reply:
I want to use it, but the `func (z *ApplicationCallTxnFields) MsgIsZero() bool` method doesn't check for nilness on the slices -- it just checks for zero length. I don't see a reason to not be strict about that, since we should always be decoding `nil` on the wire if those fields are supposed to be empty.

As for the `nil` check... I think I would rather panic in that situation than decide if a `nil` pointer is empty.